Fix cairocffi build

avnik opened the pull request:
###### Motivation for this change

cairocffi doesn't build, preventing qtile from build.
(patch doesn't applying correctly)

Patch was refreshed, as well as more packages required for tests added.

/cc @FRidh 
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

avnik committed:
pythonPackages.cairocffi: support 1.0 and 0.9

This patch ensures that Python2 can be used if cairocffi is used.
Version 1.0 dropped Python 2 support.

FRidh commented:
@GrahamcOfBorg python2.pkgs.cairocffi python3.pkgs.cairocffi

avnik commented:
On Tue, Mar 05, 2019 at 12:29:50AM -0800, Frederik Rietdijk wrote:
> @GrahamcOfBorg python2.pkgs.cairocffi python3.pkgs.cairocffi
> 

FYI cairocffi is 3.5+ library now.
Probably I should add assert as well.

FRidh commented:
@GrahamcOfBorg build python3.pkgs.cairocffi

avnik commented:
On Tue, Mar 05, 2019 at 01:10:25AM -0800, Frederik Rietdijk wrote:

Added ``disabled = pythonOlder "3.5"`` stanza

Ma27 commented:
@FRidh @avnik I'd propose to add an `cairoffi_1` package as well (similar to `cairosvg1`) to retain Python 2 support for now. This affects `thumbor`, `graphite-web` and `cairosvg1` and I think it's too early to drop Python 2 support if some dependants still need it.

Also wondering if the Python update script should contain some functionality to actually test the package before bumping it (and pushing this to master).

avnik commented:
@Ma27 ``cairocffi_py2`` be better name, than ``_1``.

Other option -- add some backporing patch to this package, I am ok with both solutions.

Ma27 commented:
> cairocffi_py2 be better name, than _1.

Ack. Didn't look at the package's version when writing this :sweat_smile: 

> Other option -- add some backporing patch to this package, I am ok with both solutions.

Can you please be more precise in what you mean? Not sure if I understand correctly what you meant :)

avnik commented:
On Thu, Mar 07, 2019 at 08:37:44PM +0000, Maximilian Bosch wrote:

> > Other option -- add some backporing patch to this package, I am ok with both solutions.
> 
> Can you please be more precise in what you mean? Not sure if I understand correctly what you meant :)
> 

I mean, if you take commit, which removing 2.x support, make reversing
patch, and will support it ;)

If you want overtake cairoffi, you are welcomed -- I care about it,
because it transitive dependency of qtile.

Ma27 commented:
> I mean, if you take commit, which removing 2.x support, make reversing patch, and will support it ;)

I doubt that it'll be that easy. Python 2.x and 3.x diverge pretty much.

But I'll have a look at this and see how to proceed here :)

Ma27 committed:
python3Packages.qasm2image: require Python 3.5 at least

According to `setup.py` in 0.8 older Python wouldn't work.

avnik commented:
On Sat, Mar 09, 2019 at 06:07:44AM -0800, Maximilian Bosch wrote:

> > I mean, if you take commit, which removing 2.x support, make reversing patch, and will support it ;)
> 
> I doubt that it'll be that easy. Python 2.x and 3.x diverge pretty much.
> 

I know. I am in middle of porting mid size project from 2.x and gevent
to 3.x and aio. Lot of pain and sorrows ;)

Ma27 commented:
Just a short update: I created a `cairocffi_0_9` package locally and played around with it. I'll push this tomorrow to this PRs branch as I'm fairly sleepy atm :)

Ma27 commented:
I pushed three commits on top of your branch:

* `8d23630`: `qasm2image` doesn't support [Python older than 3.5](https://github.com/nelimee/qasm2image/blob/2c01756946ba9782973359dbd7bbf6651af6bee5/setup.py#L128). It's more obvious to raise an error since this package doesn't support old Python rather than breaking the evaluation when trying to use the `cairocffi` derivation.

* `ad550de`: move the entire `cairocffi` build instructions into a `generic` function that is invoked for cairocffi 1.0 and 0.9. The older version should be supported for now to keep older software like thumbor working. I'd probably revert that before 19.09 as [Python 2.x will be dropped by the end of the  year](https://pythonclock.org/).

  This affects the following packages:
  * `cairosvg1`: Built with `cairocffi` 1.x by default, but when using Python 2.x (which is still supported here), `cairocffi` can be overridden by `cairocffi_0_9`. (Also wondering if we should add a `cairosvg1_py2` to make this more obvious.
   * `thumbor`: Python 3.x support is in progress, but until then we can only build with Python 2 here.
   * `graphite-api` with Python 2.x (`nix-build -E 'with import ./. {}; pythonPackages.graphite_api.override { cairocffi = pythonPackages.cairocffi_0_9; }'`)
   * `graphite_web` (doesn't work with Python 2.x)
   * and probably some more...

* `da86a8e`: unbreak `thumbor` (by using `cairocffi_0_9` and adding some minor fixes to the test suite).

@FRidh @avnik it would be great if you could review this as well :)

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1311,7 +1311,7 @@ in {
   canmatrix = callPackage ../development/python-modules/canmatrix {};
 
-  cairocffi = callPackage ../development/python-modules/cairocffi {};
+  inherit (callPackage ../development/python-modules/cairocffi {}) cairocffi cairocffi_0_9;
```
How about
```suggestion
  cairocffi = if isPy3k then
    (callPackage ../development/python-modules/cairocffi {}).cairocffi
  else
    (callPackage ../development/python-modules/cairocffi {}).cairocffi_0_9;
```

Ma27 replied:
You're right! That made the fix for `thumbor` far easier. I also exposed `cairocffi_1_0` and `cairocffi_0_9` for now as it seems to be a common pattern to add all versioned attributes to the package set and alias one of those to the default package name.

dotlambda replied:
I'd prefer not to expose them because `pythonPackages` should only contain a single version of each package. Btw, I think the prettiest way to do this is by using `let inherit (calPackage ...) cairocffi cairocffi_0_9; in if isPy3k ...`.

dotlambda reviewed (changes requested):
The commit messages for cairocffi should say `pythonPackages.cairocffi: ...`.

Ma27 committed:
pythonPackages.thumbor: 6.6.0 -> 6.7.0

This patch ensures that the currently broken `thumbor`[1] package builds
and works again.

The following problems were fixed:

* Rather than placing required packages (like `gifsicle` or `exiftool`)
  into the build input list, we reference them explicitly where needed
  to ensure that the package works after the build without further
  installs.

* Skip the `test_redeye_applied` test case which is broken for a while
  now.

[1] https://hydra.nixos.org/build/90290998

Commit:
python3.pkgs.cairosvg: 2.2.1 -> 2.3.0

dotlambda commented:
Cherry picked in ab89b6b45b4e79b3d8b62a8533a6eaf51606f48a..3274b1de0fbf9229efc2af26e07cf7281421dcc1.

Btw, I also pushed f721655455fb1868eb483d76873f56a53aa93a38 (cherry picked in cdc714fe4a0f7f136de03eceffd9bbcad1a8b560).

kamilchm commented:
Is there a way to use it with overridden `withXcffib` like in https://github.com/NixOS/nixpkgs/blob/6004e5dda3da46716cebcff3898fe9c877929b1f/pkgs/applications/window-managers/qtile/default.nix#L4 ?
I can't make it work after this change.

dotlambda commented:
@kamilchm Sorry for messing that up. An easy way to make it work would be to use three files: `default.nix` and `0_9.nix`, which both call `common.nix`. I'll make the change once I have time for it.
